rfcs/rfc-007-dynamic-program-dispatch.md

```diff
@@ -0,0 +1,183 @@
+
+```diff
+diff --git a/common/src/account.rs b/common/src/account.rs
+index 5abd642..89290ea 100644
+--- a/common/src/account.rs
++++ b/common/src/account.rs
+@@ -11,6 +11,12 @@ pub struct Account {
+     pub userdata: Vec<u8>,
+     /// contract id this contract belongs to
+     pub program_id: Pubkey,
++
++    /// this account contains a program (and is strictly read-only)
++    pub executable: bool,
++
++    /// the loader for this program (Pubkey::default() for no loader)
++    pub loader_program_id: Pubkey,
+ }
+
+ impl Account {
+@@ -19,6 +25,8 @@ impl Account {
+             tokens,
+             userdata: vec![0u8; space],
+             program_id,
++            false,
++            Pubkey::default(),
+         }
+     }
+ }
+```
+
+
+The executable flag may only be set by the `program_id` the account belongs to.
+Once the executable flag is set:
+1. The bank enforces the account userdata read-only even by `program_id`
+2. The executable flag cannot be unset
+3. The account may be used as input to `SystemProgram::Spawn` to forge a new `program_id`
+
+The `loader_program_id` field is set by `SystemProgram::Spawn` and is used by
+bank to locate the required loader during transaction dispatch.
+
+### Common Loader Interface
+
+```rust
+//! Instructions that must be implemented by all Loaders
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub enum LoaderInstruction {
+    /// Write program data into an account
+    ///
+    /// * key[0] - the account to write into.
+    ///
+    /// The transaction must be signed by key[0]
+    Write { offset: u32, prog: Vec<u8> },
+
+    /// Finalize an account loaded with program data for execution.
+    /// The exact preparation steps is loader specific but on success the loader must set the executable
+    /// bit of the Account
+    ///
+    /// * key[0] - the account to prepare for execution
+    ///
+    /// The transaction must be signed by key[0]
+    Finalize,
+}
+```
+
+`LoaderInstruction` is only used for transactions where the account at
+`Transaction::key[N]` is not executable.
+
+The loader also includes the *program interpreter* within it.  On transactions
+where the `Transaction::key[N]` account is executable and has a
+`loader_program_id`, the loader instead invokes the program in
+`Transaction::key[N]` passing along the transaction userdata and only
+`Transaction::key[1..N-1]`.
+
+
+### SystemProgram additions
+
+```rust
+    /// Transforms `executable_account_id` into a valid `program_id` for future transactions.
+    /// This transaction must be signed by `executable_account_id`.
+    ///
+    /// executable_account_id: An account with its `executable` flag set and non-empty `loader_program_id()`
+    ///
+    /// On success:
+    /// * The current `program_id` of `executable_account_id` will be saved in `loader_program_id`
+    /// * The `program_id` of `executable_account_id` will be set to `executable_account_id`
+    ///
+    SystemProgram::Spawn{
```

Comment:
What is the purpose of Spawn? shouldn't the loader assign `program_id` and `loader_program_id` once the new program is "Finalized"

Reply:
`Spawn` is the "syscall magic" needed to assign `program_id` and `loader_program_id` in a very controlled way.  If the loader (ie, any program) could just update those fields then they could escape their `program_id` and potentially gain write access to accounts owned by other programs in the system.

`Spawn` ensures that before an account turns into a `program_id`:
* the contents of the account have been vouched for by the loader (ie, the executable flag is set)
* the account owner wants this to happen (the transaction signature)
* The `loader_program_id` cannot be set to something arbitrary, it can only be set to the loader that vouched for it.